### ISSUE TITLE
Include symbols and SourceLink with packages

### ISSUE THREAD

**Issue by douglasg14b** [NRules/NRules#332]
Symbol files are required for [SourceLink](https://github.com/dotnet/sourcelink/blob/main/README.md) which makes using the library considerably more pleasant since users can navigate through the package source directly in their IDE, including debugging & stepping through it. Makes exploration, investigation, and discovery much MUCH easier and pleasant.

It's also required for any sort of usable debugging and error investigation.

MS Docs on it: https://learn.microsoft.com/en-us/nuget/create-packages/symbol-packages-snupkg

You just need to add the following to your `.csproj`:

```xml
<PropertyGroup>
    <IncludeSymbols>true</IncludeSymbols>
    <SymbolPackageFormat>snupkg</SymbolPackageFormat>
</PropertyGroup>
```

And then when you build, you get a `.snupkg` output, which you upload the nuget like any other package. 

Having this for the latest version would be wonderful. Huge developer experience boon with negligible time cost, and no downsides.

**Comment by douglasg14b** [comment]
@snikolayev Note that this doesn't necessarily have to be a milestone, new version release, sort of things. You can retroactively add symbol packages to every previous build, including the current v0.9.3 as long as the release build is reproducible.

OFC that would be a bit of a slog, which is why I'd recommend just doing it for the currently released version :)

**Comment by snikolayev** [comment]
@douglasg14b I understand. But I'd rather treat releases as immutable - not going to go back and publish symbol packages for old ones

**Comment by snikolayev** [comment]
After doing more research on this, there is more depth to this issue.

NRules already includes pdb files in the nuget packages, but due to this bug (https://github.com/dotnet/sdk/issues/1458) they are not copied to the output folder when NRules is used via a PackageReference by an end-user application. This is actually fixed in .net SDK 7 and can be enabled via project flags.

However, it seems that the debugging experience can still be improved by distributing symbols differently (see below) w/o the need to wait on the new SDK, nor does this seem like the mainstream way to distribute debug symbols anymore. Also, the integrity and traceability of the artifacts can be improved by using SourceLink.

According to this nuget issue (https://github.com/NuGet/docs.microsoft.com-nuget/issues/2343) there are several decisions that need to be made when setting up debug symbols distribution:
1. PDB format - a) windows or b) portable. 
2. PDB distribution model. a) In a nupkg file, alongside dll files, relying on build tooling to propagate them to the output folder of the user application. b) in a separate snupkg file, relying on a symbol server support from the nuget repository. c) embed symbols inside the dlls, so that they are trivially available in the output folder of the application
3. Embed source repository linkage to the package, so that it can be traced to sources and dev tooling can use that to show actual source information when debugging. 

For the PDB format, the obvious choice is 1.b (portable), which is already the case with NRules and is the pre-requisite for the rest of the debug experience improvements.

For the distribution model, there are multiple tradeoffs. The main ones are additional size of dlls with embedded symbols vs the guaranteed presence of symbols if they are embedded, and not having to rely on any tooling to have the symbols available when you actually need them.
Choosing 2.c (embedded) for the following reasons:
* Symbols are always available w/o support from any tooling
* Symbols are always in sync with the binaries
* No extra build steps, or extra artifacts to manage/publish
* Size overhead is minimal (e.g. before NRules.dll 190Kb, NRules.pdb 86Kb; after NRules.dll 229Kb). So, in fact, the total package size is going to be smaller (as currently it has DLLs + PDBs), and the DLL size increase is fairly small

SourceLink seems strictly beneficial, as this also improves supply chain traceability, as the package can be linked to a specific commit.

To summarize, going to create debug symbols in a portable pdb format, embed them inside the DLLs, distribute as normal nupkg packages, with SourceLink information embedded in the package.

**Comment by douglasg14b** [comment]
Oh wow, nice followup!

I've been doing it the easy way by generating a separate `snupkg` output using default settings (Only the csproj change in the OP), and uploading it to nuget as a 2nd upload each time I update a package.

It seems to do all the above, the only headache being uploading the same package twice 🤷‍♂️ 

<hr>

Side note, the project I was using NRules in got deprecated :( 
Not entirely dead yet, but you won't see much of me unless I can get it revived, or if I integrate it into any of my personal projects.
